### ISSUE TITLE
Boolean operations don't handle "self-touching" paths

### ISSUE THREAD

**Issue by lehni** [paperjs/paper.js#874]
As much as I hate to open another can of worms, here's the deal, currently (from https://github.com/paperjs/paper.js/issues/871#issuecomment-167750108):

The current implementation of `#resolveCrossing()` is unable to correctly deal with self-overlapping paths / compound-paths, a fact that was so far hidden by another bug:

The implementation of `Path#reduce()` was too aggressive in its treatment of consecutive collinear straight curves. The idea was to merge multiple following straight lines that go in the same direction to one line, and `reduce()` did that, but it also removed lines that go 180° against the previous / next line, hence changing the actual geometry.

Not filtering these out unveiled this new issue, as can be seen in on the [boolean-fix branch]( https://github.com/paperjs/paper.js/commits/boolean-fix branch) branch after e3e9ee75e193fab2ad130abdefbb879c4185448c

Since most of the encountered paths where this is an issue are very complex, I am going to attempt to craft some simple geometries that cause the same problems. More to come, but so far I have no solution ready for this : /

**Comment by iconexperience** [comment]
One problem that I can see is that we are currently deleting paths if the area is zero. 

In the following very simple example I could think of dividing the path at the marked point into two paths, one rectangle, and one path consisting of two lines.

![image](https://cloud.githubusercontent.com/assets/2122773/12036005/584de966-ae42-11e5-85ef-26eb09f7c2b6.png)

But since we are deleting paths with zero area, only the rectangle will remain, so we will get the same result as with the current algorithm.

**Comment by lehni** [comment]
So here finally a first failing example:

Let's start with [a simple shape](http://sketch.paperjs.org/#S/VU8xDoMwDPxKlAmkKAJGUKd+oCojYUjBhQgwKAlULeLvTYg61MPJPlt3552inIDmtBzANj1ltJlbP29Sk0XaPiUXgvAiN9fzOzRWYjdCtAskrpZZoc1JlSYJIw5qFnijPuDozNMOaoFHXAgUeEryp9LGXle9AW/VplqI/DbsFBrQNspYsPUG0U8nZiETN9BNgNZUac3PDH/q6zi+SxhdWGhdfKtXKNxnDw1yOK8Nzav6+AI=) with a "self-touching" geometry (meaning the contour is partly touching itself).

If I partly overlap another rectangle over it and create a compound path with both items, I end up with [a complex shape](http://sketch.paperjs.org/#S/jZCxbsMgEIZfBbGESBaKLWVx1ckvEDWj7cE11wQFHxZgV0mUd+9hag+dynDi7hf//x1Pjt0AvOTnG4T+yjPeWxX7uXNs7MI1Z+8M4Zud6C4/oA8dXgyIZ4OMzmg1hpLV+eGQMSptluZeP4DGRRxTaRt87d8abHCxlF/a+VBNbgap9KwViKgmTaMHF0SRpdgYIFaffZaYpIfLABh8nbdyYUjuK3TxL+hjZD7+Zd5W2ZhX11/Tyg6jnVBFc7HgJKpi3YHWM6ayxjp6sbOOwmG3aZMx9zMYggJFenATkEYf/+mguy1snpd1+/oB) that `resolveCrossings()` can't resolve:

[Final Sketch](http://sketch.paperjs.org/#S/jZExb4MwEIX/isUSoiILqLJQZWLtUDUjMLhwTdyYM7INURrlv/eMA0OlSmWw4L3zu++OW4Sih6iIDmdw7SlKolZ3/nsShg3CnTK2ZwgX9kbv/B1aJ/CoIL7VyOgZtERXsCpL04TR0SRBt/IbSM69TEdT4337UmONcyT/lMa6cjQT8E5OsoPYu8GTaMG4OE9CW98gXnK2SWDiFo49oLNV1vCZIaQv0Pm/oHeeefebeR1lZV5SH6Gl7gc9YufD4xknUOXLDDSeUqVW2tCNjTbUHDarNyp1PYAiKOjId2aEtYsBS9JcN6J0YS0k0oxWOqmRPe1Z9ewJ0+Zh/R0o4cJb2hJ4jsHoLyrhonVygldxBbOmUjn9+A8D4jzvxkZF1dx/AA==)

Please note that you need to run the `boolean-fix` code to see this properly fail, for the reasons outlined above (`reduce()` masking the problem).

The result opened in Illustrator, and plucked apart a bit to unveil its inner nature, looks like this:

![screen shot 2015-12-29 at 19 35 44](https://cloud.githubusercontent.com/assets/213275/12039772/7ed8f14a-ae63-11e5-92d6-b6e54c5fb634.png)

**Comment by lehni** [comment]
And if you run your test suite with the `boolean-fix` branch, you will get dozens of new failing issues that all choke on this : (

**Comment by iconexperience** [comment]
So what parts of the code a contribute to this problem? There is certainly `Path.reduce()`, then there is `tracePaths()`(which removed paths with zero area). What other parts do we have to look at?

**Comment by lehni** [comment]
I started a write-up of what's happening, need to find time to wrap it up. `Path#reduce()` is fixed on `boolean-fix`branch, that's what unmasked these problems.

**Comment by lehni** [comment]
So here's the write-up, part 1

This issue may require us to rethink the resolving of self-intersections.

But first a description of the different players and their role in the current process of handling boolean operations:

`getIntersections() / getCrossings()`:

Methods to determine the intersections between two paths, or the intersection of a path with itself (if no 2nd argument is provided). The difference between the two is that `getIntersections()` returns locations that barely touch also, while `getCrossings()` only returns actual crossings, meaning the path has to cross the other path. If the method's parameter `includeOverlaps` is set to true, it also returns so called overlaps, where the two paths are on top of each other for more than just a point, and it is unclear without further inspection if the overlap ends up being a a crossing or a *touching*.

`divideLocations()`:

Receives a list of locations and divides the involved path-items at the given locations, meaning it adds new segments at the given location without actually splitting the path into separate bits yet.

If any of the involved locations is the result of a call to `getIntersections() / getCrossings()`, then the newly added segment, or the existing segment, if the intersection happens to be on one that was there before, is linked up with the intersection object on the other path, and vice versa. This is then used later to walk path chains and cross easily at intersections / crossings.

The method also handles multi intersections in the same location, by creating linked lists of intersections through their hidden `_prev / _next` properties.

`isValid()`:

This method is used to determine if a given segment is to be considered part of the final boolean result, or if it can be discarded. It does this by looking at the winding contribution number of a given segment, taking into account the chosen boolean operator. For reason of simplicity, I am not explaining the adjust parameter yet.

In case of resolving of self-intersections, `isValid()` is currently not doing anything clever. It simply assumes that whenever a crossing is encountered, we can switch over to the other path. And as long as no paths are *self-touching*, the resulting closed paths are indeed not self-intersecting... (but when they do touch, we will end up in trouble)

`tracePaths()`:

This is where the main boolean operation magic happens. For reason of simplicity, I am first assuming that in any crossing there is only one possible
next step if we do cross, and am not talking about `findBestIntersection()` and the two passes in which it is called yet. The method performs these steps:

1. Loop through the segments until a valid first segment is found, using `isValid()` calls.
2. Loop through the segments that follow the start segment, by calling `getNext()` on each segment. 
3. If any segment has an interseection, use `isValid()` along with some other checks (e.g. is the other semgent already visited?) to determine if we can switch to the other path. Note that this really happens thorugh `findBestIntersection()`, but more on that later.
4. Add copies of the encountered segments to the resulting path and mark the original segments as visited. This is why we don't need to split the orignal paths: We keep adding copies as walk along the original paths.
5. Stop if we reach the starting segment or encounter an already visited segment with no other way to go:
	- If we managed to close the path, add it to the final result
	- If this results in an open path, something is wrong and the code currently complains about it through `console.error()`.

`resolveCrossings()`:

This method is applied to both paths / compound-paths involved in the boolean operation before the actual operation. It involves this sequence of steps:

1. Use `getCrossings()` to find self-intersections on the path, and divide it in each intersection through `divideLocations()`. This means that there is now guaranteed to be a segment in each intersection, and the segment is linked up with the intersection on the path.
2. Call `tracePaths()` to create new paths that don't self-intersect.
3. The paths are then reoriented and sorted based on their winding contribution and fill rule. This used to be `reorient()` but is now integrated in `resolveCrossings()`
4. Return a path / compound-path that is guaranteed to be non-intersecting, and of which all paths and sub-paths have a coherent orientation that we can work with in the boolean operations.

---

This is how far I got. What's missing is a description of the `propagateWinding()` method, but there are explanations about it in issue #875, and  the main `computeBoolean()` function, but this should be straight forward after all the explanations above, as most of the magic happens in the places outlined above. Also missing is a description of the handling of multiple intersections in one location, through `findBestIntersection()` and the `adjusted` parameter in `isValid()`.  These things will have to wait until 2016.

**Comment by lehni** [comment]
@iconexperience I hope the text above helps with understanding the core of boolean operations better. I think what we need to do is figure out a way to reliably split a self-intersecting path that also has self-touching parts into bits that don't self-intersect anymore. As you can see in my example in https://github.com/paperjs/paper.js/issues/874#issuecomment-167850010, the current approach leads to wrong results because we blindly cross at any intersection. I am not sure how we can easily determine where to cross and not cross though, since we can't rely on winding numbers. Self-touching regions won't change their winding contribution, and currently we don't even have to calculate winding numbers in the `resolveCrossings()` calls, only after.

**Comment by iconexperience** [comment]
@lehni Thanks for the great writeup, this is really helpful. Let's try to add some of this as comments to the code.

Just a question to make sure I understand what the problem is. Is the following example one of the cases that you are talking about? The path does not cross in the center, it only touches.

![image](https://cloud.githubusercontent.com/assets/2122773/12070861/1e48a0e6-b08e-11e5-8474-dd9ffe47a841.png)

<a href="http://sketch.paperjs.org/#S/hY9BCsIwEEWvMmSlEKSpdJPiHYQu2yxiOlppmpQkrUjp3U0WolDEgRn4j8fAX4iRAxJOqh6D6gglyrYpz9LByOAEBh9wlqHbLR5vA5rged24xkCcmmUZhXgE/aAiouKb5Fsp30jHHxL794q9rUQEBaWtx5ZDcBOu+zLRtCM7XCetnxVqVAHb2CwZZWx8cSj70d5jNcJrsb4A">Here is the sketch</a>

**Comment by lehni** [comment]
No, I am talking about touching in more than just a point, along a curve (what we call an overlap in the curve handling, but I'm not sure that terminology helps here, since we're not talking about overlapping shapes).

In my example above, it's the line that goes to the center of the rectangle and back on top of itself.

The overlap detection code in the curve handling currently doesn't handle such "self-touching" curves within a path, simply because of one line of code:

https://github.com/paperjs/paper.js/blob/develop/src/path/Curve.js#L1717

If I remove the check for `!param.startConnected && !param.endConnected` then these overlaps get already detected. And this makes sense. Looking at the example again, you'll see that the overlap range share their start / end point and so far were not considered worth checking, but they need to be checked too.

My plan now is to improve the overlap handling in the curve code, and store more information about the overlaps. The code also needs to detect overlaps along more than one curve properly, and store where it starts and ends. The question is also, in the scenario above, are there two overlaps (the line going to the center, and the one going back), or is it just one overlap, since the two are connected, consisting of both lines.

During `resolveCrossings()`, we can then either better decide if we should cross in case of such overlaps or not, or if that doesn't work well, introduce a new first step in which we remove these if two overlaps actually touch in their start / end point (or were merged to one, see above). Once the proper information about overlaps is known, this removal is actually quite easy.

I think one of these two strategies will work. I prefer the first, but am not sure it's possible yet. In either scenario,the line will be gone in the resulting shape, which is fine by me.

**Comment by lehni** [comment]
And while we're at it, perhaps we can find a better word for these than 'overlap', given its ambiguity? I have no suggestion yet though...

**Comment by iconexperience** [comment]
Regarding the term *overlap*, I think we took it from this paper:

  http://web.mit.edu/hyperbook/Patrikalakis-Maekawa-Cho/node111.html

So it appears to be the correct term. But I will look out for an alternative.

**Comment by lehni** [comment]
Maybe it's fine. As long as it's clear it's about curves, not paths / shapes.

**Comment by iconexperience** [comment]
Well, how about calling it *curveOverlap*? That should make it clear and it does not hurt, because after minification the code will have the same size.

**Comment by iconexperience** [comment]
I think that I am beginning to understand.

1. What we desperately need is a segment point at the beginning and end of the path overlap (there goes the term *curveOverlap*). If you look at the following example (the middle part is supposed to be overlapping), the start and end of the overlap has only a segment on one side, the other is missing. These segments need to be added before we can decide what to do with the overlaps. This means that handling the overlaps must happen after `divideLocations()`

![image](https://cloud.githubusercontent.com/assets/2122773/12071824/995f688c-b0c0-11e5-8519-94968f9af9f9.png)

2. After we identified the beginning and end of each overlap, we can use the angle between the tangents to decide if the overlap is a crossing or only a touching. If the angle changes direction, we have a crossing. This is a totally untested idea, but it might work.

![image](https://cloud.githubusercontent.com/assets/2122773/12071856/9fdf963a-b0c2-11e5-8e91-526bae684aba.png)

**Comment by iconexperience** [comment]
@lehni I would like to know what you want to achieve finally. As you know, currently all paths with area zero will be removed, so all overlap regions will be removed anyway. Would it be alright for you if the paths looked as follows after `preparePath()`?

![image](https://cloud.githubusercontent.com/assets/2122773/12071980/bc63c190-b0c8-11e5-8f22-ede5dca5aeed.png)

Or do you want to keep the overlaps, because if a stroke is applied, the paths certainly look different without the overlaps.

**Comment by lehni** [comment]
I compared with Adobe Illustrator and it also removes these. I think I'm fine with those gone, it's just that currently we don't remove them well. But I have a couple of ideas already as to what to do. I think it can be solved rather quickly!

**Comment by iconexperience** [comment]
Have you tried what happens if you simply remove all the overlapping parts of the path? If you make sure that both parts of the overlap have a segments at the start and end of the overlap, then you can simply combine the segments at the beginning to one and do the same with the two segments at the end (and the ones inbetween). Shouldn't this already solve the whole issue?

**Comment by iconexperience** [comment]
Of course my previous suggestions would not work in some cases. For example if you take a rectangle and scale it to 0 in y direction, then the result of the preparation would be no path at all. I am sure we do not want this.

**Comment by iconexperience** [comment]
As mentioned in #882, there can be cases of self touching paths for which we currently do not detect any overlaps. This can happen with curves that have handles on the line between the end points but not between the end points. A curve of this type may look like this:

![image](https://cloud.githubusercontent.com/assets/2122773/12074614/9766fe08-b15d-11e5-8d8d-6e95619de1c5.png)

<a href="http://sketch.paperjs.org/#S/fZBNDsIgEEavMmHVRtqUmm4wHsCdicu2i0rH1ojQAGqM8e5CtQt/IgEmvHnD4rsR1RyRcLI5oBM9oUToNrzPjYGBwRIUXmDduD66WeyOqJzlUFamUvBaJcsyCv6q6Rue/8H5F2Z/MCu+cPHEFMLxOxnrh5Vnk5W8tF8Wy97+qqdmTUFIbbHl4MwJ7/Fi6oQ6sHR3kvK6QYnCYeuTCtboCK2slphK3UUV2fvIgENFYOYTTTt0K+XQWD+292IUpxJV5/p44dPfGmwOgw4zhJf1/QE=">Here is the sketch for a complete closed self-touching path</a> where the overlap is not found.

![image](https://cloud.githubusercontent.com/assets/2122773/12074599/fd5d7f1c-b15c-11e5-85f8-57bef22d4a53.png)

**Comment by lehni** [comment]
So now some more about `reduce()`. First a description of how it currently works.

If we take the above *problematic* shape again, and move the first and second segments further to the right, to create a situation where not the full range of the two involved curves touch, I thought we end up in a situation where the current implementation of a too aggressive `reduce()` might produce a wrong shape, but it actually doesn't: 

[Sketch](http://sketch.paperjs.org/#S/bZDBbsIwDIZfJcqpFVFokXYp4rQXmMaxzSFrXIgoLkrSThvi3XHSDk2CHKzEv/37i68c9Rl4xfcnCO2RC94OJr4n7dhFh2PJdgzhm33QXX5CGzQeesiuDTI6l8FiqFhdFoVgFJSY897+AqU3MU1BNXjLtw02GG2HrvMQyLd+I7lQSUizZGedD++jm0AaO1kDWWybNYseXMg2YuaJk7O/AbmYYaWHwxkw+LpUMsHl/9wfYrGIbLVbaLZPNeWrmgfn2Pc/e+hpH2DoJ8GNkOT1egFxYMY24tNKvxzoUzLzvKrV7Q4=)

Here's how the shape looks before `reduce()`:

<img width="245" alt="screen shot 2016-01-04 at 11 17 42" src="https://cloud.githubusercontent.com/assets/213275/12087629/d535ca0c-b2d4-11e5-82b3-140454d36ed8.png">

And here after:

<img width="229" alt="screen shot 2016-01-04 at 11 17 46" src="https://cloud.githubusercontent.com/assets/213275/12087632/d93c0f9e-b2d4-11e5-8ed0-fb62b77361af.png">

As you can see, it does exactly what we expect from the overlap removal code. But this shouldn't happen in `reduce()`, and is just a side effect of code that was supposed to *merge* multiple consecutive lines with same direction in a path, e.g. something like this path:

<img width="302" alt="screen shot 2016-01-04 at 11 20 05" src="https://cloud.githubusercontent.com/assets/213275/12087673/302ab2e2-b2d5-11e5-95e6-6a55a36069aa.png">

Turning into this:

<img width="322" alt="screen shot 2016-01-04 at 11 20 13" src="https://cloud.githubusercontent.com/assets/213275/12087676/367988ee-b2d5-11e5-916d-f045813e5e3a.png">

So without intending to, we gave `reduce()` powers that helped solve many boolean operation edge cases. And in a way that was good, because it gave us the confidence that it can be solved. But now it's time to get this right, properly :)

Now I thought that given the simple implementation of `Path#reduce()`, there must be situations where it's doing the wrong thing. So I tried reversing the above path, thinking that it would remove the longer curve first, and hence move a segment. But it does the right thing again!

[Reversed Sketch](http://sketch.paperjs.org/#S/bVBBboMwEPyK5ROolgWReiHKqR+ImiP44OIlsUKWaG2o2ih/7xrSCKn1YWXP7M6M9ybRXkBW8nCG2J6kku3g0nuyJK42nkqxEwifYs93/Q5ttHjsIbs1KPhcB4+xEnVZFEpwMWrBg/8GhjcJ5mIavOfbBhtMskPXBYisW78yXZiZmL105ynEt5Em0M5P3kGWxhbOYwCK2UYteZJz9muQqyWsDnC8AMZQl0bP4fKV+pMsHqR42T3SbP/0lP/1PHOOff91gJ73AY5/EmmEFU0wAYV1eAI3tgngDX8Q2POsHWRVm/sP)

I now understand why this is happening: It is because when a curve is removed, only the second segment of the curve is actually removed, while its value of `handleOut` is move to the first segment. If you step through the above scenarios in a debugger you will see how that ends up doing the right thing.

**Comment by lehni** [comment]
And those two issues were created by d6a8538674e6c638640d93b695c754e2a09104a8 and bf65866b58b0193aac778c72de7585dfe79ecbd3, which in turn have fixed #870. Catch-22
